Automattic/jetpack
# Docker: use `.env` files for customisation

1) Adds possibility to use `.env`-files to customize containers.

   The setup allows easy switching of PHP version, WP version, XDebug vars etc — none of these is implemented in this PR though.

    The `.env` file will be git-ignored and survives between WordPress reinstalls (`yarn docker:install`). There is `default.env` which is used by default but any values in `.env` override those. You could override only e.g. domain and password settings and not define anything else.

   In typical situation `.env` file would be available in the same directory where `docker-compose` command is run. 

   In such situation variables would be usable already in `docker-compose.yml` file itself. Because we don't run `docker-compose` from the same folder where our `.env` file is, we end up not having them available there.

    We do have them available _inside_ our containers because in `docker-compose.yml` we define also `env_file` which reads variables and passes them on to containers:
    ```
    env_file:
      - default.env
      - .env
    ```
   Defining anything using `environment` key in `docker-compose.yml` would override anything defined in those `.env`-files.

    https://docs.docker.com/compose/environment-variables/

    https://docs.docker.com/compose/compose-file/#env_file

2) Move scripts from `yarn docker:install` and `yarn docker:multisite-convert` to separate bash files in `./bin` directory. 

    Improves OS (i.e. Windows) support since scripts are run inside the container. 

    Previoulsy these install scripts would break the site easily because of domain-config issues. This change solves those troubles. 

## Testing

- Clean up your previous setup: `yarn docker:clean`
- Copy configs: `cp ./docker/default.env ./docker/.env` and modify them. Leave some lines out from your file to let them be picked from `default.env` file.
- Run `yarn docker:up`
- Ensure containers are running and inspect their configuration: `yarn docker:compose config` — you should see your config options there.
- Ensure you have `WP_DOMAIN` set and run `yarn docker:install` and `yarn docker:multisite-convert` and ensure your site works.
- Ensure you can connect to SFTP container


---

Taken out from https://github.com/Automattic/jetpack/pull/9116
